system/codesigner/gpgcodesigner.cpp

```diff
@@ -0,0 +1,250 @@
+#include "jlib.hpp"
+#include "jexcept.hpp"
+#include "jlog.hpp"
+#include "gpgcodesigner.hpp"
+#include "atomic"
+
+class GpgCodeSigner : implements ICodeSigner
+{
+    virtual bool sign(const char * text, const char * userId, const char * passphrase, StringBuffer & errmsg, StringBuffer & signedText) override;
+    virtual bool verifySignature(const char * text, StringBuffer & signer, StringBuffer & errmsg) override;
+    virtual bool hasSignature(const char * text) const override;
+    virtual StringBuffer &stripSignature(const char * text, StringBuffer & unsignedText) const override;
+    virtual StringArray &getUserIds(StringArray & userIds, StringBuffer & errmsg) override;
+private:
+    void initGpg(void);
+    bool getKeyGrip(const char * user, StringBuffer & keygrip);
+    void clearPassphrase(const char * key);
+    bool isGpgV1 = false;
```

Comment:
ultra picky: this should also be atomic to avoid and tiny race condition on weak ordering processors.  (Another thread may see the assignment to isInitialised before the update to isGPgV1)

Reply:
done